src/os/bluestore/NVMEDevice.cc

```diff
@@ -931,12 +988,8 @@ int NVMEDevice::read_random(uint64_t off, uint64_t len, char *buf, bool buffered
   Task *t = new Task(this, IOCommand::READ_COMMAND, aligned_off, aligned_len, 1);
   int r = 0;
   t->ctx = &ioc;
-  t->fill_cb = [buf, t, off, len]() {
-    t->copy_to_buf(buf, off-t->offset, len);
-  };
 
-  ++ioc.num_pending;
-  ioc.nvme_task_first = t;
+  read_split(this, aligned_off, &ioc, buf, aligned_len, t, off, len);
```

Comment:
why not consolidate the 
```c++
Task *t = new Task(t...)
```

into `read_split()`as a single loop? something like (not tested):

```c++
auto begin = aligned_off;
const auto aligned_end = begin + aligned_len;
const auto orig_end = orig_off + orig_len;
Task* primary = nullptr;
for (; begin < aligned_end; begin += split_size) {
  auto to_read = min(split_size, aligned_end - begin);
  auto task = new Task(dev,  IOCommand::READ_COMMAND, begin, to_read,
    primary);
  auto output_off = primary ? 0 : orig_off - aligned_off;
  auto output_len = min(orig_end - begin);
  task->fill_cb = [=] {
    task->copy_to_buf(output_buf, output_off, output_len);
  }
  task->ctx = ioc;
  if (!primary) {
    primary = task;
  }
  append_task(ioc, t);
}
```

Reply:
For random_read and read, if the I/O is splitted, primary task will not be executed. Since the completion of subtasks are completed in out of order, better to allocate the primary task out and not put into the split function.

Reply:
I am afraid that your single loop is not exactly correct.

If the io is not split, there is no primary task, we do not want to allocate a primary task when the io does not need split.
If the io is need split, there will be primary task (this task will not be executed). Suppose there are 4 subtasks, task1, task2, task3, task4, their completion can be out of order. So you'd better not to re-use one task as primary task, this will contain bugs.

Reply:
Change the while loop into for loop.

Reply:
> If the io is not split, there is no primary task, we do not want to allocate a primary task when the io does not need split.

 what i suggest is to remove `Task *t = new Task(this, IOCommand::READ_COMMAND, off, len, 1);` call and move it into the `read_split()` call instead of duplicating it.

Reply:
>  better to allocate the primary task out side the function and not put into the split function.

the split function's name might be misleading in this case, could rename it to "make_read_tasks()"

Reply:
> > If the io is not split, there is no primary task, we do not want to allocate a primary task when the io does not need split.
> 
> what i suggest is to remove `Task *t = new Task(this, IOCommand::READ_COMMAND, off, len, 1);` call and move it into the `read_split()` call instead of duplicating it.

Better not to do that, since if we do it in read_split, then we need to return Task * as return value, but for aio_read, we do not need that, and there is Task t used to check the return_code (after the I/O is done)

Reply:
> then we need to return Task * as return value

i don't see the reason to do so. as task and subtasks (if any) are connected and addressed using `ioc->nvme_task_first` and `ioc->nvme_task_last` already. why do we need to "return Task *"?